File: gremlin-go/driver/pdtRegistry.go
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package gremlingo

import "reflect"

// PDTAdapter defines how to hydrate/dehydrate a provider-defined type.
type PDTAdapter struct {
	TypeName   string
	FromFields func(map[string]interface{}) (interface{}, error)
	ToFields   func(interface{}) (map[string]interface{}, error)
}

// PrimitivePDTAdapter defines how to hydrate/dehydrate a primitive provider-defined type.
type PrimitivePDTAdapter struct {
	TypeName   string
	FromString func(string) (interface{}, error)
	ToString   func(interface{}) (string, error)
}

// PDTRegistry maps type names to their hydration adapters.
type PDTRegistry struct {
	adaptersByName          map[string]*PDTAdapter
	adaptersByType          map[reflect.Type]*PDTAdapter
	primitiveAdaptersByName map[string]*PrimitivePDTAdapter
	primitiveAdaptersByType map[reflect.Type]*PrimitivePDTAdapter
}

// NewPDTRegistry creates an empty PDTRegistry.
func NewPDTRegistry() *PDTRegistry {
	return &PDTRegistry{
		adaptersByName:          make(map[string]*PDTAdapter),
		adaptersByType:          make(map[reflect.Type]*PDTAdapter),
		primitiveAdaptersByName: make(map[string]*PrimitivePDTAdapter),
		primitiveAdaptersByType: make(map[reflect.Type]*PrimitivePDTAdapter),
	}
}

// RegisterFuncs registers hydration/dehydration functions for a type name.
func (r *PDTRegistry) RegisterFuncs(typeName string, fromFields func(map[string]interface{}) (interface{}, error), toFields func(interface{}) (map[string]interface{}, error)) {
	adapter := &PDTAdapter{TypeName: typeName, FromFields: fromFields, ToFields: toFields}
	r.adaptersByName[typeName] = adapter
}

// RegisterFuncsWithType registers hydration/dehydration functions for a type name and associates a Go type for dehydration lookup.
func (r *PDTRegistry) RegisterFuncsWithType(typeName string, targetType reflect.Type, fromFields func(map[string]interface{}) (interface{}, error), toFields func(interface{}) (map[string]interface{}, error)) {
	adapter := &PDTAdapter{TypeName: typeName, FromFields: fromFields, ToFields: toFields}
	r.adaptersByName[typeName] = adapter
	r.adaptersByType[targetType] = adapter
}

// RegisterType registers a struct type for reflection-based hydration using "pdt" struct tags.
func (r *PDTRegistry) RegisterType(typeName string, targetType reflect.Type) {
	r.adaptersByName[typeName] = &PDTAdapter{
		TypeName: typeName,
		FromFields: func(fields map[string]interface{}) (interface{}, error) {
			obj := reflect.New(targetType).Elem()
			for i := 0; i < targetType.NumField(); i++ {
				field := targetType.Field(i)
				tag := field.Tag.Get("pdt")
				if tag == "" {
					tag = field.Name
				}
				if val, ok := fields[tag]; ok && val != nil {
					obj.Field(i).Set(reflect.ValueOf(val))
				}
			}
			return obj.Interface(), nil
		},
	}
}

// GetAdapterByType returns the adapter registered for the given Go type, or nil.
func (r *PDTRegistry) GetAdapterByType(t reflect.Type) *PDTAdapter {
	return r.adaptersByType[t]
}

// Hydrate converts a CompositePDT into a domain object using the registered adapter.
// Returns the raw PDT if no adapter is found or if hydration fails.
// Nested registered PDTs in Fields are always hydrated recursively, even when the outer has no adapter.
func (r *PDTRegistry) Hydrate(pdt *CompositePDT) interface{} {
	if pdt == nil {
		return nil
	}
	hydratedFields := make(map[string]interface{}, len(pdt.Fields))
	for k, v := range pdt.Fields {
		if nested, ok := v.(*CompositePDT); ok {
			hydratedFields[k] = r.Hydrate(nested)
		} else if nested, ok := v.(*PrimitivePDT); ok {
			hydratedFields[k] = r.HydratePrimitive(nested)
		} else {
			hydratedFields[k] = v
		}
	}
	adapter, ok := r.adaptersByName[pdt.Name]
	if !ok {
		return &CompositePDT{Name: pdt.Name, Fields: hydratedFields}
	}
	result, err := adapter.FromFields(hydratedFields)
	if err != nil {
		return pdt
	}
	return result
}

// RegisterPrimitiveFuncs registers hydration/dehydration functions for a primitive type name.
func (r *PDTRegistry) RegisterPrimitiveFuncs(typeName string, fromString func(string) (interface{}, error), toString func(interface{}) (string, error)) {
	adapter := &PrimitivePDTAdapter{TypeName: typeName, FromString: fromString, ToString: toString}
	r.primitiveAdaptersByName[typeName] = adapter
}

// RegisterPrimitiveFuncsWithType registers hydration/dehydration functions for a primitive type name
// and associates a Go type for dehydration lookup.
func (r *PDTRegistry) RegisterPrimitiveFuncsWithType(typeName string, targetType reflect.Type, fromString func(string) (interface{}, error), toString func(interface{}) (string, error)) {
	adapter := &PrimitivePDTAdapter{TypeName: typeName, FromString: fromString, ToString: toString}
	r.primitiveAdaptersByName[typeName] = adapter
	r.primitiveAdaptersByType[targetType] = adapter
}

// GetPrimitiveAdapterByType returns the primitive adapter registered for the given Go type, or nil.
func (r *PDTRegistry) GetPrimitiveAdapterByType(t reflect.Type) *PrimitivePDTAdapter {
	return r.primitiveAdaptersByType[t]
}

// HydratePrimitive converts a PrimitivePDT into a domain object using the registered primitive adapter.
// Returns the raw PDT if no adapter is found or if hydration fails.
func (r *PDTRegistry) HydratePrimitive(pdt *PrimitivePDT) interface{} {
	if pdt == nil {
		return nil
	}
	adapter, ok := r.primitiveAdaptersByName[pdt.Name]
	if !ok {
		return pdt
	}
	result, err := adapter.FromString(pdt.Value)
	if err != nil {
		return pdt
	}
	return result
}

File: gremlin-js/gremlin-javascript/lib/structure/PDTRegistry.ts
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

import { CompositePDT, PrimitivePDT } from './graph.js';

export interface PdtAdapter {
  serialize: (obj: any) => Record<string, any>;
  deserialize: (fields: Record<string, any>) => any;
}

export interface PrimitivePdtAdapter {
  toValue: (obj: any) => string;
  fromValue: (value: string) => any;
}

/**
 * A standalone registry that allows users to register adapters for hydrating
 * raw {@link CompositePDT} and {@link PrimitivePDT} instances
 * into domain-specific objects.
 */
export class PDTRegistry {
  private readonly _adapters: Map<string, PdtAdapter> = new Map();
  private readonly _adaptersByClass: Map<Function, { typeName: string; adapter: PdtAdapter }> = new Map();
  private readonly _primitiveAdapters: Map<string, PrimitivePdtAdapter> = new Map();
  private readonly _primitiveAdaptersByClass: Map<Function, { typeName: string; adapter: PrimitivePdtAdapter }> = new Map();

  register(typeName: string, adapter: PdtAdapter, targetClass?: Function): void {
    this._adapters.set(typeName, adapter);
    if (targetClass) {
      this._adaptersByClass.set(targetClass, { typeName, adapter });
    }
  }

  registerPrimitive(typeName: string, adapter: PrimitivePdtAdapter, targetClass?: Function): void {
    this._primitiveAdapters.set(typeName, adapter);
    if (targetClass) {
      this._primitiveAdaptersByClass.set(targetClass, { typeName, adapter });
    }
  }

  hydrate(pdt: any): any {
    if (!(pdt instanceof CompositePDT)) return pdt;
    const adapter = this._adapters.get(pdt.name);
    const hydratedFields: Record<string, any> = {};
    let changed = false;
    for (const [k, v] of Object.entries(pdt.fields)) {
      if (v instanceof CompositePDT) {
        const h = this.hydrate(v);
        hydratedFields[k] = h;
        if (h !== v) changed = true;
      } else if (v instanceof PrimitivePDT) {
        const h = this.hydratePrimitive(v);
        hydratedFields[k] = h;
        if (h !== v) changed = true;
      } else {
        hydratedFields[k] = v;
      }
    }
    if (!adapter) {
      return changed ? new CompositePDT(pdt.name, hydratedFields) : pdt;
    }
    try {
      return adapter.deserialize(hydratedFields);
    } catch (e: any) {
      console.warn(`PDT hydration failed for '${pdt.name}': ${e.message}`);
      return pdt;
    }
  }

  hydratePrimitive(pdt: any): any {
    if (!(pdt instanceof PrimitivePDT)) return pdt;
    const adapter = this._primitiveAdapters.get(pdt.name);
    if (!adapter) return pdt;
    try {
      return adapter.fromValue(pdt.value);
    } catch (e: any) {
      console.warn(`Primitive PDT hydration failed for '${pdt.name}': ${e.message}`);
      return pdt;
    }
  }

  hasAdapter(typeName: string): boolean {
    return this._adapters.has(typeName);
  }

  hasPrimitiveAdapter(typeName: string): boolean {
    return this._primitiveAdapters.has(typeName);
  }

  getSerializer(typeName: string): ((obj: any) => Record<string, any>) | null {
    const adapter = this._adapters.get(typeName);
    return adapter ? adapter.serialize : null;
  }

  getAdapterByClass(cls: Function): { typeName: string; serialize: (obj: any) => Record<string, any> } | null {
    const entry = this._adaptersByClass.get(cls);
    if (!entry) return null;
    return { typeName: entry.typeName, serialize: entry.adapter.serialize };
  }

  getPrimitiveAdapterByClass(cls: Function): { typeName: string; toValue: (obj: any) => string } | null {
    const entry = this._primitiveAdaptersByClass.get(cls);
    if (!entry) return null;
    return { typeName: entry.typeName, toValue: entry.adapter.toValue };
  }
}
